aten/src/ATen/native/ReduceOps.cpp

```diff
@@ -599,6 +599,109 @@ Tensor& diff_out(const Tensor& self, int64_t n, int64_t dim, const c10::optional
+    } else {
+      auto ax_dx = spacing[i];
+      auto center  = (at::slice(self,direction, 2   ) - at::slice(self, direction, 0, -2 ) ) / ax_dx;
+      auto prepend = (at::slice(self,direction, 1, 2) - at::slice(self, direction, 0, 1  ) ) / ax_dx;
+      auto append  = (at::slice(self,direction, -1  ) - at::slice(self, direction, -2, -1) ) / ax_dx;
+      result.push_back(prepend_append_on_dim(center/2, prepend, append, direction));
+    }
+  }
+  return result;
+}
+
+std::vector<Tensor> gradient_helper_scalar(const Tensor& self, ArrayRef<Scalar> spacing, IntArrayRef axis, int64_t edge_order) {
+  pre_check_gradient(self, axis, spacing.size(), edge_order);
+  std::vector<Tensor> result;
+  for (const auto i : c10::irange(axis.size())) {
+      int64_t direction = maybe_wrap_dim(axis[i], self.dim());
+      auto ax_dx = spacing[i];
+      auto center  = (at::slice(self,direction, 2   ) - at::slice(self, direction, 0, -2 ) ) / ax_dx;
+      auto prepend = (at::slice(self,direction, 1, 2) - at::slice(self, direction, 0, 1  ) ) / ax_dx  ;
+      auto append  = (at::slice(self,direction, -1  ) - at::slice(self, direction, -2, -1) ) / ax_dx ;
+      result.push_back(prepend_append_on_dim(center/2, prepend, append, direction));
+  }
+  return result;
+}
+
+std::vector<int64_t> gradient_axis_preprocess(const Tensor& self, c10::optional<int64_t> dim) {
+  std::vector<int64_t> axis;
+  if (!dim.has_value()) {
+    for (const auto i : c10::irange(self.dim())) {
+      axis.push_back(i);
+    }
+  }
+  else {
+    axis.push_back(dim.value());
+  }
+  return axis;
+}
+
+std::vector<Tensor> gradient(const Tensor& self, TensorList spacing, IntArrayRef dim, int64_t edge_order) {
+    return gradient_helper(self, spacing, dim, edge_order);
+}
+
+std::vector<Tensor> gradient(const Tensor& self, TensorList spacing, c10::optional<int64_t> dim, int64_t edge_order) {
+  return gradient_helper(self, spacing, IntArrayRef(gradient_axis_preprocess(self, dim)), edge_order);
+}
+
+std::vector<Tensor> gradient(const Tensor& self, ArrayRef<Scalar> spacing, IntArrayRef dim, int64_t edge_order) {
+  return gradient_helper_scalar(self, spacing, dim, edge_order);
+}
+
+std::vector<Tensor> gradient(const Tensor& self, ArrayRef<Scalar> spacing, c10::optional<int64_t> dim, int64_t edge_order) {
+  return gradient_helper_scalar(self, spacing, IntArrayRef(gradient_axis_preprocess(self, dim)), edge_order);
+}
+
+std::vector<Tensor> gradient(const Tensor& self, const Scalar& unit_size, IntArrayRef dim, int64_t edge_order) {
+  std::vector<Scalar> spacing;
+  for(const auto i : c10::irange(dim.size())) {
+    spacing.push_back(unit_size);
+  }
+  return gradient_helper_scalar(self, spacing, dim, edge_order);
+}
+
+std::vector<Tensor> gradient(const Tensor& self, const c10::optional<Scalar>& unit_size, c10::optional<int64_t> dim, int64_t edge_order) {
+  std::vector<Scalar> spacing;
+  for(const auto i : c10::irange(dim.has_value() ? 1 : self.dim())) {
```

Comment:
See other vector comments for an idea of how to tweak these range-based for loops